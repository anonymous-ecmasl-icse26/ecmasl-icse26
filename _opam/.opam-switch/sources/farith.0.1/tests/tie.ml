open Farith2

(** not tested *)

let tiebreak a b n =
  begin
    (* Tie breaks at [2^(n-1) + e] with [n] bits precision *)
    let m = n-1 in
    let up = Q.mul_2exp Q.one m in
    let q = Q.(up + Q.make (Z.of_int a) (Z.of_int b)) in
    let f0 = F.of_qint ~bits:(n+1) q in (* exact *)
    let f1,f2 = F.seize ~bits:n f0 in
    Format.printf "------------------------------------------@\n" ;
    Format.printf "         - %a    (%a)@\n" F.pp f1 F.pp f1 ;
    Format.printf "1p%d+%d/%d = %a (%a)@\n" m a b pp f0 F.pp f0 ;
    Format.printf "         + %a    (%a)@\n" pp f2 F.pp f2 ;
    Format.printf "mantissa =  %a@\n" Z.pp_print (Z.shift_left Z.one n) ;
    let f1 = F.neg f0 in
    let pos_ne = F.round ~mode:F.NE ~bits:n f0 in
    let pos_na = F.round ~mode:F.NA ~bits:n f0 in
    let neg_ne = F.round ~mode:F.NE ~bits:n f1 in
    let neg_na = F.round ~mode:F.NA ~bits:n f1 in
    Format.printf "+NE %a (%a)@\n" F.pp pos_ne pp pos_ne ;
    Format.printf "-NE %a (%a)@\n" F.pp neg_ne pp neg_ne ;
    Format.printf "+NA %a (%a)@\n" F.pp pos_na pp pos_na ;
    Format.printf "-NA %a (%a)@\n" F.pp neg_na pp neg_na ;
    Format.print_flush () ;
  end

let () =
  begin
    tiebreak 1 2 F.b64 ;
    tiebreak 3 2 F.b64 ;
  end
