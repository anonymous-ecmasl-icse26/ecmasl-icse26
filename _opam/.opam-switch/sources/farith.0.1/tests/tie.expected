------------------------------------------
         - +4503599627370496    (+1p52)
1p52+1/2 = +9007199254740993p-1 (+9007199254740993p-1)
         + +4503599627370497    (+4503599627370497)
mantissa =  9007199254740992
=NE +1p52
+NE +1p52 (+4503599627370496)
-NE -1p52 (-4503599627370496)
=NA +4503599627370497p0
+NA +4503599627370497 (+4503599627370497)
-NA -4503599627370497 (-4503599627370497)
------------------------------------------
         - +4503599627370497    (+4503599627370497)
1p52+3/2 = +9007199254740995p-1 (+9007199254740995p-1)
         + +4503599627370498    (+2251799813685249p1)
mantissa =  9007199254740992
=NE +2251799813685249p1
+NE +2251799813685249p1 (+4503599627370498)
-NE -2251799813685249p1 (-4503599627370498)
=NA +2251799813685249p1
+NA +2251799813685249p1 (+4503599627370498)
-NA -2251799813685249p1 (-4503599627370498)
